src/main/java/org/mars/rover/kata/MarsNavigator.java

```diff
@@ -1,60 +1,72 @@
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Scanner;
+import java.util.stream.IntStream;
 
 public class MarsNavigator {
     CommandSet providedInput;
 
-    ArrayList<ArrayList<Coordinate>> grid;
+    Grid grid;
 
     List<MarsRover> marsRovers;
 
     public MarsNavigator(CommandSet providedInput) {
         this.providedInput = providedInput;
-        this.grid = new ArrayList<>();
+        this.grid = Grid.createGrid(providedInput.gridX(), providedInput.gridY());
         this.marsRovers = new ArrayList<>();
     }
 
-
     public void processCommandSet() {
-        this.createGrid();
-        this.marsRovers = this.constructRovers(this.providedInput.roverInstructions()); // todo write test
-    }
-
-    protected void createGrid() {
-        int width = this.providedInput.gridX();
-        int height = this.providedInput.gridY();
-
-        for (int i = 0; i < width; i++) {
-            ArrayList<Coordinate> row = new ArrayList<>();
-
-            for (int j = 0; j < height; j++) {
-                row.add(new Coordinate(i, j));
-            }
-
-            grid.add(row);
-        }
+        this.marsRovers = this.constructRovers(this.providedInput.roverInstructions());
+        this.processRoverCommands(this.providedInput.roverInstructions());
     }
 
     public List<MarsRover> constructRovers(List<RoverInstructions> roverInstructions) {
         return roverInstructions.stream()
                 .map(RoverInstructions::initialPosition)
                 .map(initialPosition -> new MarsRover(
-                        initialPosition.x(),
-                        initialPosition.y(),
+                        initialPosition.coordinate().x(),
+                        initialPosition.coordinate().y(),
                         initialPosition.direction()
                 ))
                 .toList();
     }
 
+    public void processRoverCommands(List<RoverInstructions> roverInstructions) {
+        IntStream.range(0, this.marsRovers.size())
+                .forEach(index -> {
+                    var currentRover = this.marsRovers.get(index);
+                    var instructionsForThisRover = roverInstructions.get(index);
+
+                    instructionsForThisRover.roverCommands().forEach(
+                            instruction -> {
+                                currentRover.setPosition(
+                                        this.wrapEdges(instruction.execute(currentRover.getPosition()))
+                                );
+                            }
+                    );
+                });
+    }
+
+    protected Position wrapEdges(Position newPosition) {
+        int numRows = providedInput.gridX();
+        int numCols = providedInput.gridY();
+
+        int wrappedX = (newPosition.coordinate().x() + numRows) % numRows;
+        int wrappedY = (newPosition.coordinate().y() + numCols) % numCols;
+
+        return newPosition.withCoordinate(new Coordinate(wrappedX, wrappedY));
```

Comment:
Bardzo czytelna jest ta funkcja, świetnie to się czyta. Myślę, że nawet osoba nie do końca zaznajomiona z programowaniem byłaby w stanie to koncepcyjnie zrozumieć.